### DISCUSSION TITLE
A question about 'dFscaling'

### DISCUSSION THREAD

**Discussion by 101709** [mdolab/dafoam#716] [General]
1. I wrote a function that added a Hanning window to achieve a weighted average, and achieved fast and accurate convergence on the weighted average of the objective function. But I don't know how `unsteadyObjFuncsScaling_`’should be modified.
![image](https://github.com/user-attachments/assets/55a2e44b-3be1-4d1a-b4db-cb3d6fc0e207)

2. In the first version, I used the following Settings directly, meaning that `dFScaling` was the same for each time step
![image](https://github.com/user-attachments/assets/307bd713-1052-4c2c-bc2d-584f8673f116)

3. However, later I realized that the proportion of each time step should be different, so I made the following changes, which I set in `pyDAfoam.py`, because the function in `DAsolver.C` will only be called once during initialization, and the result obtained is a little different from that of the first version including direct time average, I wonder if this modification is correct because calculation of the total derivative is different from the unmodified version. I don't know how to verify what the true gradient is.
![image](https://github.com/user-attachments/assets/a8053567-7540-48fe-8007-49b6f4e9c7fe)

4. Also, if n is not in the range of `objFuncStartTimeIndex` and `objFuncEndTimeIndex`, `dFScaling = 0.0`, I understand that outside of this range, because according to the formula, the partial derivative of `F` with respect to `x` and the partial derivative of `F` with respect to `w` are both 0, so the total derivative of the objective function with respect to the design variable should be 0, so, I've set the cut-off time for the reverse mode to `objFuncEndTimeIndex`, and as before, This calculation of the total derivative is also different from the unmodified version. This is the result on a cylinder.
![image](https://github.com/user-attachments/assets/495e2c85-21b1-44c8-bb6f-69b4cad4c43c)
The `base` curve shown in the figure above is dafoam analysis without any modification.  `change1` is obtained according to the modification of the fourth point, meaning that the reverse derivative goes from 3s to 1s instead of 3s to 0s. 
```
        "objFuncTimeOperator": "average",
        "objFuncStartTime": 1.0,
        "objFuncEndTime": 3.0,
```
And I saw from the output that when `dFScaling=0`, the total derivative of the corresponding time step is not 0, which is not what I understood.

I do not know whether there is a problem in the understanding of `dFscaling`, so I would like to ask you. My question may be a little long, thank you very much for your patience and reply. I also hope that this windowing function can be helpful to the subsequent development.

**Comment by 101709** [comment]
This is the code I changed.
[code.zip](https://github.com/user-attachments/files/18083673/code.zip)

**Comment by friedenhe** [comment]
dFScaling is used for unsteady objective functions, such as summing and averaging an objective over a time period. Say your objFunc is F_avg = sum(f1+f2+f3+...fN) / N, in the adjoint computation, the adjoint vector psi should be computed as (dRdW)^T * psi1 = df1/dw / N. So 1/N is the dFScaling. In the screenshots you shared, I feel that 2 and 3 are duplicated. If you have computed unsteadyObjFuncScaling_ in DASolver.C, you don't have to compute it again in pyDAFoam.py, you can just get its value by calling self.solver.getObjFuncUnsteadyScaling

**Reply by 101709** [reply]
Picture 2 and picture 3 are different. The `dFScaling` of the former is constant and is the same for each time step. The latter `dFScaling` is different for each time step. In addition, what is your understanding of my question 4 above. Why does this difference happen?
![image](https://github.com/user-attachments/assets/c8e8b173-2947-43b0-b725-02a7fcc88011)
The red is what I changed.

**Comment by friedenhe** [comment]
@101709, we will release a new version v4, which will have an easier interface to handle time-dependent dFScaling. At this moment, I am not quite sure what caused the difference in your Q4..
